File: AspNetCore/MCRCOSucursales/AspNetCore/Models/Rentals/McrcoSucursalesModel.cs
//McrcoSucursalesModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.OData.ModelBuilder;
using Microsoft.EntityFrameworkCore;
using Cxc.TablasBasicas.Models.v1;

namespace MilesCarRental.Rentals.Models.v1
{
    [Table("Sucursales", Schema = "MCRCO")]
    public class McrcoSucursales : IDisposable
	{
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int McrcoSucursalesId {get; set;}

        //[Column(TypeName = "Date")]
        public DateTime McrcoSucursalesFechaInicio {get; set;}

        [MaxLength(105)]
        public string McrcoSucursalesDescripcion {get; set;}

        [MaxLength(250)]
        public string McrcoSucursalesDireccion {get; set;}

        public int CiudadDepartamentoId {get; set;}

        public int Ciudadid {get; set;}

        [NotMapped]
        public string CntCiudadesComp
        {
        	get { return $"{CiudadDepartamentoId}/{Ciudadid}"; }
        	private set {}
        }
        [Column(TypeName = "decimal(9, 6)")]
        public decimal McrcoSucursalesLongitud {get; set;}

        [Column(TypeName = "decimal(8, 5)")]
        public decimal McrcoSucursalesLatitud {get; set;}

        public Enum_MCRCOSucursalesEstado McrcoSucursalesEstado {get; set;}

        public CntCiudades CntCiudades { get; set; }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        private bool disposed = false;
        
        protected virtual void Dispose(bool disposing)
        {
            if(!this.disposed)
            {
                disposed = true;
            }
        }
	}

    #region Enums
    public enum Enum_MCRCOSucursalesEstado : int
    {
        Activo = 1,
        En_Mantenimiento = 2,
        Cerrado = 3
    }
    #endregion


    /// <summary>
    /// Extensión para registrar mapping con el Entity Framework y oData
    /// </summary>
	public static class McrcoSucursalesExtension
	{
        #region EF Mapping
        public static ModelBuilder McrcoSucursalesMapping(this ModelBuilder modelBuilder)
        {
            var entity = modelBuilder.Entity<McrcoSucursales>();

            //PrimaryKey
            entity.HasKey(c => new { c.McrcoSucursalesId });

           //Relationships
            entity.HasOne(typeof(CntCiudades), "CntCiudades")
                .WithMany()
                .HasForeignKey("CiudadDepartamentoId", "Ciudadid")
                .OnDelete(DeleteBehavior.Restrict); // no ON DELETE


			return modelBuilder;
		}

        /// <summary>
        /// Mapping con el oData Framework
        /// </summary>
        /// <param name="oDataModelBuilder"></param>
        public static void McrcoSucursalesMapping(this ODataConventionModelBuilder oDataModelBuilder)
        {
            var entityConfig = oDataModelBuilder.EntitySet<McrcoSucursales>(nameof(McrcoSucursales));

            var entity = entityConfig.EntityType;

            // PrimaryKey
            entity.HasKey(c => new { c.McrcoSucursalesId });



            //Ignored properties for oData
            // entityConfig.EntityType.Ignore(x => x.Summary);
        }
        #endregion
    }
}

File: Angular/sucursales-mcr/src/app/rentals/mcrco.sucursales/mcrco.sucursales.model.ts
export enum EnumMcrcoSucursalesEstado {
  'Activo' = '1',
  'En_Mantenimiento' = '2',
  'Cerrado' = '3'
}

export class McrcoSucursalesModel {
    public McrcoSucursalesId: number = 0;
    public McrcoSucursalesFechaInicio: Date = new Date();
    public McrcoSucursalesDescripcion: string;
    public McrcoSucursalesDireccion: string;
    public CiudadDepartamentoId: number;
    public Ciudadid: number;
    public CntCiudadesComp: string;
    public McrcoSucursalesLongitud: number = 0.0;
    public McrcoSucursalesLatitud: number = 0.0;
    public McrcoSucursalesEstado: string = 'Activo';
    public CntCiudades?: any = {};
    public _secuencia?: number = 0;
    public _estado?: string = 'N';
    public _id?: string = '';
    public _v?: number = 0;

    constructor(json?: any) {
        if(json) {
            this.McrcoSucursalesId = json.McrcoSucursalesId;
            this.McrcoSucursalesFechaInicio = json.McrcoSucursalesFechaInicio;
            this.McrcoSucursalesDescripcion = json.McrcoSucursalesDescripcion;
            this.McrcoSucursalesDireccion = json.McrcoSucursalesDireccion;
            this.CiudadDepartamentoId = json.CiudadDepartamentoId;
            this.Ciudadid = json.Ciudadid;
            this.CntCiudadesComp =  json.CntCiudadesComp;
            this.McrcoSucursalesLongitud = json.McrcoSucursalesLongitud;
            this.McrcoSucursalesLatitud = json.McrcoSucursalesLatitud;
            this.McrcoSucursalesEstado = json.McrcoSucursalesEstado;
            this.CntCiudades = json.CntCiudades;
        }
    }

    static clone(row: McrcoSucursalesModel): McrcoSucursalesModel {
        const rowCloned = Object.assign({}, row);

        delete rowCloned._secuencia;
        delete rowCloned._estado;
        delete rowCloned._id;
        delete rowCloned._v;
        delete rowCloned.CntCiudades;

        return rowCloned;
    }

    static cloneExcel(data: McrcoSucursalesModel[]): any[] {
       const dataExcel: any[] = [];
       data.forEach(dato => {
           const registro = {

              McrcoSucursalesId: dato.McrcoSucursalesId,
              McrcoSucursalesFechaInicio: dato.McrcoSucursalesFechaInicio,
              McrcoSucursalesDescripcion: dato.McrcoSucursalesDescripcion,
              McrcoSucursalesDireccion: dato.McrcoSucursalesDireccion,
              CiudadDepartamentoId: dato.CiudadDepartamentoId,
              Ciudadid: dato.Ciudadid,
              CntCiudadesComp:  dato.CntCiudadesComp,
              McrcoSucursalesLongitud: dato.McrcoSucursalesLongitud,
              McrcoSucursalesLatitud: dato.McrcoSucursalesLatitud,
              McrcoSucursalesEstado: dato.McrcoSucursalesEstado,
              CntCiudades:  dato.CntCiudades

            };
            dataExcel.push(registro);
       });
       return dataExcel;
    }

    toClipboard(separator: string = '\t'): string {
        let result = '';

        result += `${separator}${this.McrcoSucursalesId}`;
        result += `${separator}${this.McrcoSucursalesFechaInicio}`;
        result += `${separator}${this.McrcoSucursalesDescripcion}`;
        result += `${separator}${this.McrcoSucursalesDireccion}`;
        result += `${separator}${this.CiudadDepartamentoId}`;
        result += `${separator}${this.Ciudadid}`;
        result += `${separator}${this.McrcoSucursalesLongitud}`;
        result += `${separator}${this.McrcoSucursalesLatitud}`;
        result += `${separator}${this.McrcoSucursalesEstado}`;

        return result.substring(separator.length);
    }

    fromClipboard(value: string, separator: string = '\t'): McrcoSucursalesModel {
        const result = value.split(separator);

        this.McrcoSucursalesId = parseInt(result[0], 10);
        this.McrcoSucursalesFechaInicio = new Date(result[1]);
        this.McrcoSucursalesDescripcion = result[2];
        this.McrcoSucursalesDireccion = result[3];
        this.CiudadDepartamentoId = parseInt(result[4], 10);
        this.Ciudadid = parseInt(result[5], 10);
        this.McrcoSucursalesLongitud = parseFloat(result[6]);
        this.McrcoSucursalesLatitud = parseFloat(result[7]);
        this.McrcoSucursalesEstado = result[8];

        return this;
    }

}
